sympy/matrices/normalforms.py

```diff
@@ -1,35 +1,43 @@
 '''Functions returning normal forms of matrices'''
 
-from sympy.matrices.dense import diag, zeros
+from sympy.polys.polytools import Poly
+from sympy.polys.matrices import DomainMatrix
+from sympy.polys.matrices.normalforms import (
+        smith_normal_form as _snf,
+        invariant_factors as _invf,
+    )
 
 
-def smith_normal_form(m, domain = None):
+def _to_domain(m, domain=None):
+    """Convert Matrix to DomainMatrix"""
+    m = m.applyfunc(lambda e: e.as_expr() if isinstance(e, Poly) else e)
+    dM = DomainMatrix.from_Matrix(m)
+    domain = domain or getattr(m, 'domain', None)
+    if domain is not None:
+        dM = dM.convert_to(domain)
+    return dM
+
+
+def smith_normal_form(m, domain=None):
     '''
     Return the Smith Normal Form of a matrix `m` over the ring `domain`.
     This will only work if the ring is a principal ideal domain.
 
     Examples
     ========
 
-    >>> from sympy.polys.solvers import RawMatrix as Matrix
-    >>> from sympy.polys.domains import ZZ
+    >>> from sympy import Matrix, ZZ
     >>> from sympy.matrices.normalforms import smith_normal_form
     >>> m = Matrix([[12, 6, 4], [3, 9, 6], [2, 16, 14]])
-    >>> setattr(m, "ring", ZZ)
-    >>> print(smith_normal_form(m))
+    >>> print(smith_normal_form(m, domain=ZZ))
     Matrix([[1, 0, 0], [0, 10, 0], [0, 0, -30]])
 
     '''
-    invs = invariant_factors(m, domain=domain)
-    smf = diag(*invs)
-    n = len(invs)
-    if m.rows > n:
-        smf = smf.row_insert(m.rows, zeros(m.rows-n, m.cols))
-    elif m.cols > n:
-        smf = smf.col_insert(m.cols, zeros(m.rows, m.cols-n))
-    return smf
+    dM = _to_domain(m, domain)
```

Comment:
This can be made to warn deprecation by looking up for `.ring` attribute first, and raising the deprecation warning if there is the attribute.

Reply:
I'm not sure how that helps with anything. This PR removes the `RawMatrix` class so anyone following the documented way to use the `shmith_normal_form` function will find that their code breaks:
https://github.com/sympy/sympy/blob/d43e9dc5aed4d2b695a3176cdf8ad4e3c79f0e5b/sympy/matrices/normalforms.py#L14-L20
It isn't possible to make `RawMatrix` work and it does need to be completely removed from the codebase in order to establish the internal invariant of `Matrix` that its elements are `Expr`.

The `RawMatrix` class which subclasses `Matrix` and uses unsympifiable domain elements as the elements of the matrix is completely broken. It is a previous attempt to make something like `DomainMatrix` but is based on a flawed design. It isn't really possible to make something that will emulate `RawMatrix` based on `DomainMatrix` because a `RawMatrix` is constructed directly with domain elements or Poly but no domain is supplied and the domain is instead attached as the `ring` attribute.

Anyone who is currently using the `smith_normal_form` function needs to change their code in order to either use `Matrix` with `Expr` or `DomainMatrix` with domain elements and a specified domain. This PR makes both of those possible but I don't see any good way to make that work in a backwards compatible way.

Reply:
Maybe RawMatrix itself should be kept as a deprecating class for now.

Is there a reason that RawMatrix should be removed immediately for now? Maybe you want to change the Matrix internals.

Reply:
If we keep `RawMatrix` in any way then it will continue to constrain what can be done to improve `Matrix`. There are other classes that need to be changed or removed as well though like `NewMatrix` and `PolyMatrix` so until that is done we do not need to remove `RawMatrix`.

We need to establish the invariant that the matrix elements are Expr which will improve the matrices code. It also makes it possible to use DomainMatrix internally in Matrix.

Reply:
I've added back a hacky version of RawMatrix that converts all the elements to Expr but manages to pass the previous tests for the Smith normal form. It's hard to be sure that it is really compatible with current RawMatrix just because RawMatrix is so poorly defined.

Reply:
I don't think RawMatrix is needed anywhere. There was an issue with matrices having Poly elements (https://github.com/sympy/sympy/issues/15086) but that was resolved by means of PolyMatrix. That should also have been used in normalforms instead of RawMatrix but it was only being developed at the time.

Reply:
> I don't think RawMatrix is needed anywhere.

Agreed. This PR removes the only place in the sympy codebase that still uses RawMatrix. I guess it is reasonable to leave a DeprecationWarning though at least until the other issues are resolved given that it is currently the documented way to compute the Smith normal form.

The other issues are:

1. NewMatrix needs to be removed from holonomic which should just use DomainMatrix instead.
2. PolyMatrix also subclasses Matrix and changes the internal elements. It should be reimplemented on top of DomainMatrix instead.

Once those issues are also solved then removing RawMatrix would lead to the actual beneficial result that Matrix can constrain its entries to Expr and then DomainMatrix can be used internally by Matrix.